src/mesh/mesh.cpp

```diff
@@ -903,7 +906,12 @@ void Mesh::EnrollBndryFncts_(ApplicationInput *app_in) {
 // \!fn void Mesh::ApplyUserWorkBeforeOutput(ParameterInput *pin)
 // \brief Apply MeshBlock::UserWorkBeforeOutput
 
-void Mesh::ApplyUserWorkBeforeOutput(ParameterInput *pin) {
+void Mesh::ApplyUserWorkBeforeOutput(Mesh *mesh, ParameterInput *pin,
+                                     SimTime const &time) {
+  // call Mesh version
+  mesh->UserMeshWorkBeforeOutput(mesh, pin, time);
+
+  // call MeshBlock version
   for (auto &pmb : block_list) {
     pmb->UserWorkBeforeOutput(pmb.get(), pin);
   }
```

Comment:
A topic for discussion might be if we want to allow for both versions to be enrolled simultaneously.
At least for for the initialization function, the general consensus was "no" because that would avoid potential race conditions with respect to the ordering of operations.

Reply:
I'm not sure how to do this, since the function pointers are assigned in different constructors, so by this point, they are both non-`nullptr`, even if neither have been overridden.

I also tried this:
```
// check that UserMeshWorkBeforeOutput and UserWorkBeforeOutput are not both enrolled
PARTHENON_REQUIRE(
      (block_list[0]->UserWorkBeforeOutput == &UserMeshWorkBeforeOutputDefault) &&
          (mesh->UserMeshWorkBeforeOutput == &UserMeshWorkBeforeOutputDefault),
      "UserMeshWorkBeforeOutput and UserWorkBeforeOutput cannot both be enrolled "
      "simultaneously!");
```
but the comparison operator for `std::function` is deleted.

Reply:
I think you could mirror what's done, e.g., for `ProblemGenerator` or `PostInit`, in `meshblock.cpp`
```c++
  if (app_in->ProblemGenerator != nullptr) {
    ProblemGenerator = app_in->ProblemGenerator;
    // Only set default block pgen when no mesh pgen is set
  } else if (app_in->MeshProblemGenerator == nullptr) {
    ProblemGenerator = &ProblemGeneratorDefault;
  }
  if (app_in->PostInitialization != nullptr) {
    PostInitialization = app_in->PostInitialization;
    // Only set default post-init when no mesh post-init is set
  } else if (app_in->MeshPostInitialization == nullptr) {
    PostInitialization = &PostInitializationDefault;
  }
```
i.e., only set the default `MeshBlock` one if there's no `Mesh` version.
Then the check for both `!= nullptr` should work.